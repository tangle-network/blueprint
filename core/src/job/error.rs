use alloc::string::String;
use core::error::Error;
use core::fmt::Display;

#[derive(Debug)]
pub struct JobError {
    pub reason: String,
}

impl<T: Into<String>> From<T> for JobError {
    fn from(value: T) -> Self {
        Self {
            reason: value.into(),
        }
    }
}

impl Display for JobError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "{reason}", reason = self.reason)
    }
}

impl Error for JobError {}
