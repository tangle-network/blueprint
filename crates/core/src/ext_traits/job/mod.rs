mod call;
pub use call::*;
mod parts;
pub use parts::*;
