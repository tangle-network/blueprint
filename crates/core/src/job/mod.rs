//! Async functions that can be used to handle jobs.
//#![doc = include_str!("../docs/handlers_intro.md")]
//!
//! Some examples of handlers:
//!
//! ```rust
//! use blueprint_sdk::Bytes;
//!
//! // Handler that immediately returns an empty `200 OK` response.
//! async fn unit_handler() {}
//!
//! // Handler that immediately returns a `200 OK` response with a plain text
//! // body.
//! async fn string_handler() -> String {
//!     "Hello, World!".to_string()
//! }
//!
//! // Handler that buffers the request body and returns it.
//! //
//! // This works because `Bytes` implements `FromRequest`
//! // and therefore can be used as an extractor.
//! //
//! // `String` and `StatusCode` both implement `IntoResponse` and
//! // therefore `Result<String, StatusCode>` also implements `IntoResponse`
//! async fn echo(body: Bytes) -> Result<String, String> {
//!     if let Ok(string) = String::from_utf8(body.to_vec()) {
//!         Ok(string)
//!     } else {
//!         Err(String::from("Invalid UTF-8"))
//!     }
//! }
//! ```
//!
//! Instead of a direct `String`, it makes sense to use intermediate error type
//! that can ultimately be converted to `Response`. This allows using `?` operator
//! in handlers. See those examples:
//!
//! * [`anyhow-error-response`][anyhow] for generic boxed errors
//! * [`error-handling`][error-handling] for application-specific detailed errors
//!
//! [anyhow]: https://github.com/tokio-rs/axum/blob/main/examples/anyhow-error-response/src/main.rs
//! [error-handling]: https://github.com/tokio-rs/axum/blob/main/examples/error-handling/src/main.rs
//#![doc = include_str!("../docs/debugging_handler_type_errors.md")]

use crate::{
    JobCall, JobResult,
    extract::{FromJobCall, FromJobCallParts},
    job_result::IntoJobResult,
};
use alloc::boxed::Box;
use core::{fmt, future::Future, marker::PhantomData, pin::Pin};
use tower::{Layer, Service, ServiceExt};

pub mod future;
mod service;

pub use self::service::JobService;

/// Trait for async functions that can be used to handle requests.
///
/// You shouldn't need to depend on this trait directly. It is automatically
/// implemented to closures of the right types.
///
/// See the [module docs](crate::handler) for more details.
///
/// # Converting `Handler`s into [`Service`]s
///
/// To convert `Handler`s into [`Service`]s you have to call either
/// [`HandlerWithoutStateExt::into_service`] or [`Handler::with_state`]:
///
/// ```
/// use blueprint_sdk::job::JobWithoutContextExt;
/// use blueprint_sdk::{Context, Job, JobCall};
/// use tower::Service;
///
/// // this handler doesn't require any state
/// async fn one() {}
/// // so it can be converted to a service with `HandlerWithoutStateExt::into_service`
/// assert_service(one.into_service());
///
/// // this handler requires a context
/// async fn two(_: Context<String>) {}
/// // so we have to provide it
/// let handler_with_state = two.with_context(String::new());
/// // which gives us a `Service`
/// assert_service(handler_with_state);
///
/// // helper to check that a value implements `Service`
/// fn assert_service<S>(service: S)
/// where
///     S: Service<JobCall>,
/// {
/// }
/// ```
//#[doc = include_str!("../docs/debugging_handler_type_errors.md")]
///
/// # Handlers that aren't functions
///
/// The `Handler` trait is also implemented for `T: IntoResponse`. That allows easily returning
/// fixed data for routes:
///
/// ```
/// use blueprint_sdk::Router;
/// use serde_json::json;
///
/// const HELLO_JOB_ID: u32 = 0;
/// const USERS_JOB_ID: u32 = 1;
///
/// let app = Router::new()
///     // respond with a fixed string
///     .route(HELLO_JOB_ID, "Hello, World!")
///     // or return some mock data
///     .route(USERS_JOB_ID, json!({ "id": 1, "username": "alice" }).to_string());
/// # let _: Router = app;
/// ```
#[diagnostic::on_unimplemented(
    note = "Consider using `#[blueprint_sdk::debug_handler]` to improve the error message"
)]
pub trait Job<T, Ctx>: Clone + Send + Sync + Sized + 'static {
    /// The type of future calling this handler returns.
    type Future: Future<Output = Option<JobResult>> + Send + 'static;

    /// Call the handler with the given request.
    fn call(self, call: JobCall, ctx: Ctx) -> Self::Future;

    /// Apply a [`tower::Layer`] to the handler.
    ///
    /// All requests to the handler will be processed by the layer's
    /// corresponding middleware.
    ///
    /// This can be used to add additional processing to a request for a single
    /// handler.
    ///
    /// Note this differs from [`routing::Router::layer`](crate::routing::Router::layer)
    /// which adds a middleware to a group of routes.
    ///
    /// If you're applying middleware that produces errors you have to handle the errors
    /// so they're converted into responses. You can learn more about doing that
    /// [here](crate::error_handling).
    ///
    /// # Example
    ///
    /// Adding the [`tower::limit::ConcurrencyLimit`] middleware to a handler
    /// can be done like so:
    ///
    /// ```rust
    /// use blueprint_sdk::{Job, Router};
    /// use tower::limit::{ConcurrencyLimit, ConcurrencyLimitLayer};
    ///
    /// async fn handler() { /* ... */
    /// }
    ///
    /// const MY_JOB_ID: u32 = 0;
    ///
    /// let layered_handler = handler.layer(ConcurrencyLimitLayer::new(64));
    /// let app = Router::new().route(MY_JOB_ID, layered_handler);
    /// # let _: Router = app;
    /// ```
    fn layer<L>(self, layer: L) -> Layered<L, Self, T, Ctx>
    where
        L: Layer<JobService<Self, T, Ctx>> + Clone,
        L::Service: Service<JobCall>,
    {
        Layered {
            layer,
            job: self,
            _marker: PhantomData,
        }
    }

    /// Convert the handler into a [`Service`] by providing the context
    fn with_context(self, ctx: Ctx) -> JobService<Self, T, Ctx> {
        JobService::new(self, ctx)
    }
}

impl<F, Fut, Res, Ctx> Job<((),), Ctx> for F
where
    F: FnOnce() -> Fut + Clone + Send + Sync + 'static,
    Fut: Future<Output = Res> + Send,
    Res: IntoJobResult,
{
    type Future = Pin<Box<dyn Future<Output = Option<JobResult>> + Send>>;

    fn call(self, _call: JobCall, _ctx: Ctx) -> Self::Future {
        Box::pin(async move { self().await.into_job_result() })
    }
}

macro_rules! impl_job {
    (
        [$($ty:ident),*], $last:ident
    ) => {
        #[allow(non_snake_case, unused_mut)]
        impl<F, Fut, Ctx, Res, M, $($ty,)* $last> Job<(M, $($ty,)* $last,), Ctx> for F
        where
            F: FnOnce($($ty,)* $last,) -> Fut + Clone + Send + Sync + 'static,
            Fut: Future<Output = Res> + Send,
            Ctx: Send + Sync + 'static,
            Res: IntoJobResult,
            $( $ty: FromJobCallParts<Ctx> + Send, )*
            $last: FromJobCall<Ctx, M> + Send,
        {
            type Future = Pin<Box<dyn Future<Output = Option<JobResult>> + Send>>;

            fn call(self, call: JobCall, state: Ctx) -> Self::Future {
                Box::pin(async move {
                    let (mut parts, body) = call.into_parts();
                    let state = &state;

                    $(
                        let $ty = match $ty::from_job_call_parts(&mut parts, state).await {
                            Ok(value) => value,
                            Err(rejection) => return rejection.into_job_result(),
                        };
                    )*

                    let req = JobCall::from_parts(parts, body);

                    let $last = match $last::from_job_call(req, state).await {
                        Ok(value) => value,
                        Err(rejection) => return rejection.into_job_result(),
                    };

                    let res = self($($ty,)* $last,).await;

                    res.into_job_result()
                })
            }
        }
    };
}

all_the_tuples!(impl_job);

mod private {
    // Marker type for `impl<T: IntoResponse> Job for T`
    #[allow(missing_debug_implementations)]
    pub enum IntoJobResultHandler {}
}

impl<T, Ctx> Job<private::IntoJobResultHandler, Ctx> for T
where
    T: IntoJobResult + Clone + Send + Sync + 'static,
{
    type Future = core::future::Ready<Option<JobResult>>;

    fn call(self, _call: JobCall, _ctx: Ctx) -> Self::Future {
        core::future::ready(self.into_job_result())
    }
}

/// A [`Service`] created from a [`Job`] by applying a Tower middleware.
///
/// Created with [`Job::layer`]. See that method for more details.
pub struct Layered<L, J, T, Ctx> {
    layer: L,
    job: J,
    _marker: PhantomData<fn() -> (T, Ctx)>,
}

impl<L, J, T, Ctx> fmt::Debug for Layered<L, J, T, Ctx>
where
    L: fmt::Debug,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Layered")
            .field("layer", &self.layer)
            .finish()
    }
}

impl<L, J, T, Ctx> Clone for Layered<L, J, T, Ctx>
where
    L: Clone,
    J: Clone,
{
    fn clone(&self) -> Self {
        Self {
            layer: self.layer.clone(),
            job: self.job.clone(),
            _marker: PhantomData,
        }
    }
}

impl<L, J, Ctx, T> Job<T, Ctx> for Layered<L, J, T, Ctx>
where
    L: Layer<JobService<J, T, Ctx>> + Clone + Send + Sync + 'static,
    L::Service: Service<JobCall> + Clone + Send + 'static,
    <L::Service as Service<JobCall>>::Response: IntoJobResult,
    <L::Service as Service<JobCall>>::Future: Send,
    J: Job<T, Ctx>,
    T: 'static,
    Ctx: 'static,
{
    type Future = future::LayeredFuture<L::Service>;

    fn call(self, req: JobCall, state: Ctx) -> Self::Future {
        use futures_util::future::{FutureExt, Map};

        let svc = self.job.with_context(state);
        let svc = self.layer.layer(svc);

        #[allow(clippy::type_complexity)]
        let future: Map<
            _,
            fn(
                Result<
                    <L::Service as Service<JobCall>>::Response,
                    <L::Service as Service<JobCall>>::Error,
                >,
            ) -> _,
        > = svc.oneshot(req).map(|result| match result {
            Ok(res) => res.into_job_result(),
            Err(_err) => todo!("JobService needs to return a result"),
        });

        future::LayeredFuture::new(future)
    }
}

/// Extension trait for [`Job`]s that don't have state.
///
/// This provides convenience methods to convert the [`Job`] into a [`Service`] or [`MakeService`].
///
/// [`MakeService`]: tower::make::MakeService
pub trait JobWithoutContextExt<T>: Job<T, ()> {
    /// Convert the handler into a [`Service`] and no state.
    fn into_service(self) -> JobService<Self, T, ()>;
}

impl<H, T> JobWithoutContextExt<T> for H
where
    H: Job<T, ()>,
{
    fn into_service(self) -> JobService<Self, T, ()> {
        self.with_context(())
    }
}
