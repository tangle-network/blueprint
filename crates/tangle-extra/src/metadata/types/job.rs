use crate::serde::new_bounded_string;
use alloc::borrow::Cow;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::bounded_collections::bounded_vec::BoundedVec;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::tangle_primitives::services::field::FieldType;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::tangle_primitives::services::jobs::JobDefinition as SubxtJobDefinition;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::tangle_primitives::services::jobs::JobMetadata as SubxtJobMetadata;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::tangle_primitives::services::types::PricingModel;

/// A Job Definition is a definition of a job that can be called.
/// It contains the input and output fields of the job with the permitted caller.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct JobDefinition<'a> {
    pub job_id: u64,
    /// The metadata of the job.
    pub metadata: JobMetadata<'a>,
    /// These are parameters that are required for this job.
    /// i.e. the input.
    pub params: Vec<FieldType>,
    /// These are the result, the return values of this job.
    /// i.e. the output.
    pub result: Vec<FieldType>,
    /// The pricing model for the job.
    pub pricing_model: PricingModel<u32, u128>,
}

impl From<SubxtJobDefinition> for JobDefinition<'static> {
    fn from(value: SubxtJobDefinition) -> Self {
        Self {
            job_id: 0,
            metadata: value.metadata.into(),
            params: value.params.0,
            result: value.result.0,
            pricing_model: value.pricing_model,
        }
    }
}

impl From<JobDefinition<'_>> for SubxtJobDefinition {
    fn from(value: JobDefinition<'_>) -> Self {
        Self {
            metadata: value.metadata.into(),
            params: BoundedVec(value.params),
            result: BoundedVec(value.result),
            pricing_model: value.pricing_model,
        }
    }
}

#[derive(Default, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct JobMetadata<'a> {
    /// The Job name.
    pub name: Cow<'a, str>,
    /// The Job description.
    pub description: Option<Cow<'a, str>>,
    /// Runtime profile from benchmarking (optional).
    /// Generated via `cargo tangle blueprint profile` command.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub profile: Option<JobProfile>,
}

/// Runtime profile of a job from benchmarking.
///
/// This is generated by running the job with sample inputs and measuring
/// actual resource usage. Benefits:
/// - `FaaS` compatibility detection
/// - VM sizing recommendations
/// - Cost estimation
/// - `QoS` monitoring baselines
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct JobProfile {
    /// Average execution time in milliseconds
    pub avg_duration_ms: u64,
    /// Peak memory usage in MB
    pub peak_memory_mb: u32,
    /// 95th percentile execution time
    pub p95_duration_ms: u64,
    /// Whether job maintains state across invocations
    pub stateful: bool,
    /// Whether job opens persistent connections
    pub persistent_connections: bool,
    /// Sample size used for profiling
    pub sample_size: u32,
}

impl From<SubxtJobMetadata> for JobMetadata<'static> {
    fn from(value: SubxtJobMetadata) -> Self {
        Self {
            name: String::from_utf8_lossy(&value.name.0.0).into_owned().into(),
            description: value
                .description
                .map(|desc| String::from_utf8_lossy(&desc.0.0).into_owned().into()),
            profile: None, // Profiles not yet stored on chain
        }
    }
}

impl From<JobMetadata<'_>> for SubxtJobMetadata {
    fn from(value: JobMetadata<'_>) -> Self {
        Self {
            name: new_bounded_string(value.name),
            description: value.description.map(new_bounded_string),
        }
    }
}

impl Default for JobDefinition<'_> {
    fn default() -> Self {
        Self {
            job_id: 0,
            metadata: JobMetadata::default(),
            params: Vec::new(),
            result: Vec::new(),
            pricing_model: PricingModel::PayOnce { amount: 0 },
        }
    }
}
