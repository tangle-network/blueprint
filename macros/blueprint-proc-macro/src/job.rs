use crate::eigenlayer::event_listener::generate_eigenlayer_event_handler;
use crate::shared::{pascal_case, type_to_field_type};
use crate::tangle::event_listener::generate_tangle_event_handler;
use gadget_blueprint_proc_macro_core::{FieldType, JobDefinition, JobMetadata, JobResultVerifier};
use proc_macro::TokenStream;
use quote::{format_ident, quote};
use std::collections::{BTreeMap, HashSet};
use syn::ext::IdentExt;
use syn::parse::{Parse, ParseStream};
use syn::{Ident, ItemFn, LitInt, LitStr, Token, Type};

// Defines custom keywords
mod kw {
    syn::custom_keyword!(id);
    syn::custom_keyword!(params);
    syn::custom_keyword!(result);
    syn::custom_keyword!(verifier);
    syn::custom_keyword!(evm);
    syn::custom_keyword!(event_handler);
    syn::custom_keyword!(protocol);
    syn::custom_keyword!(instance);
    syn::custom_keyword!(event);
    syn::custom_keyword!(event_converter);
    syn::custom_keyword!(callback);
    syn::custom_keyword!(skip_codegen);
}

pub(crate) fn job_impl(args: &JobArgs, input: &ItemFn) -> syn::Result<TokenStream> {
    // Extract function name and arguments
    let fn_name = &input.sig.ident;
    let fn_name_string = fn_name.to_string();
    let job_def_name = format_ident!("{}_JOB_DEF", fn_name_string.to_ascii_uppercase());
    let job_id_name = format_ident!("{}_JOB_ID", fn_name_string.to_ascii_uppercase());

    let syn::ReturnType::Type(_, result) = &input.sig.output else {
        return Err(syn::Error::new_spanned(
            &input.sig.output,
            "Function must have a return type of Result<T, E> where T is a tuple of the result fields",
        ));
    };

    // check that the function has a return type of Result<T, E>
    match **result {
        Type::Path(ref path) => {
            let seg = path.path.segments.last().unwrap();
            if seg.ident != "Result" {
                return Err(syn::Error::new_spanned(
                    result,
                    "Function must have a return type of Result<T, E> where T is a tuple of the result fields",
                ));
            }
        }
        _ => {
            return Err(syn::Error::new_spanned(
                result,
                "Function must have a return type of Result<T, E> where T is a tuple of the result fields",
            ));
        }
    }

    let mut param_types = BTreeMap::new();
    for input in &input.sig.inputs {
        if let syn::FnArg::Typed(arg) = input {
            if let syn::Pat::Ident(pat_ident) = &*arg.pat {
                let ident = &pat_ident.ident;
                let ty = &*arg.ty;
                let added = param_types.insert(ident.clone(), ty.clone());
                if added.is_some() {
                    return Err(syn::Error::new_spanned(
                        ident,
                        "tried to add the same field twice",
                    ));
                }
            }
        }
    }

    let job_id = &args.id;
    let params_type = args.params_to_field_types(&param_types)?;
    let result_type = args.result_to_field_types(result)?;

    let event_handler_gen = if args.skip_codegen {
        proc_macro2::TokenStream::default()
    } else {
        // Generate the Job Handler.
        generate_event_handler_for(input, args, &param_types, &params_type, &result_type)
    };

    // Extract params and result types from args
    let job_def = JobDefinition {
        metadata: JobMetadata {
            name: fn_name_string.clone().into(),
            // filled later on during the rustdoc gen.
            description: None,
        },
        params: params_type,
        result: result_type,
        verifier: match &args.verifier {
            Verifier::Evm(contract) => JobResultVerifier::Evm(contract.clone()),
            Verifier::None => JobResultVerifier::None,
        },
    };

    let job_def_str = serde_json::to_string(&job_def).map_err(|err| {
        syn::Error::new_spanned(
            input,
            format!("Failed to serialize job definition to json: {err}"),
        )
    })?;

    let gen = quote! {
        #[doc = "Job definition for the function "]
        #[doc = "[`"]
        #[doc = #fn_name_string]
        #[doc = "`]"]
        #[automatically_derived]
        #[doc(hidden)]
        pub const #job_def_name: &str = #job_def_str;

        #[doc = "Job ID for the function "]
        #[doc = "[`"]
        #[doc = #fn_name_string]
        #[doc = "`]"]
        #[automatically_derived]
        pub const #job_id_name: u8 = #job_id;

        #input

        #event_handler_gen
    };

    Ok(gen.into())
}

#[allow(clippy::too_many_lines)]
pub fn generate_event_handler_for(
    f: &ItemFn,
    job_args: &JobArgs,
    param_types: &BTreeMap<Ident, Type>,
    params: &[FieldType],
    result: &[FieldType],
) -> proc_macro2::TokenStream {
    let fn_name = &f.sig.ident;
    let fn_name_string = fn_name.to_string();
    let struct_name = format_ident!("{}EventHandler", pascal_case(&fn_name_string));
    let job_id = &job_args.id;
    let event_handler = &job_args.event_handler;

    // Get all the params names inside the param_types map
    // and not in the params list to be added to the event handler.
    let x = param_types.keys().collect::<HashSet<_>>();
    let y = job_args.params.iter().collect::<HashSet<_>>();
    let diff = x.difference(&y).collect::<Vec<_>>();
    let additional_params = diff
        .iter()
        .map(|ident| {
            let mut ty = param_types[**ident].clone();
            // remove the reference from the type and use the inner type
            if let Type::Reference(r) = ty {
                ty = *r.elem;
            }
            quote! {
                pub #ident: #ty,
            }
        })
        .collect::<Vec<_>>();

    let additional_params_in_call = diff
        .iter()
        .map(|ident| {
            let ty = &param_types[**ident];
            let (is_ref, is_ref_mut) = match ty {
                Type::Reference(r) => (true, r.mutability.is_some()),
                _ => (false, false),
            };
            if is_ref && is_ref_mut {
                quote! { &mut self.#ident, }
            } else if is_ref {
                quote! { &self.#ident, }
            } else {
                quote! { self.#ident, }
            }
        })
        .collect::<Vec<_>>();

    let params_tokens = params
        .iter()
        .enumerate()
        .map(|(i, t)| {
            let ident = format_ident!("param{i}");
            let index = syn::Index::from(i);
            match event_handler {
                EventHandlerArgs::Eigenlayer { .. } => {
                    quote! {
                        let #ident = inputs.#index;
                    }
                }
                EventHandlerArgs::Tangle => crate::tangle::field_type_to_param_token(&ident, t),
            }
        })
        .collect::<Vec<_>>();

    let fn_call_params = params
        .iter()
        .enumerate()
        .map(|(i, _)| {
            let ident = format_ident!("param{i}");
            quote! {
                #ident,
            }
        })
        .collect::<Vec<_>>();
    let fn_call = if f.sig.asyncness.is_some() {
        quote! {
            let job_result = match #fn_name(
                #(#additional_params_in_call)*
                #(#fn_call_params)*
            ).await {
                Ok(r) => r,
                Err(e) => {
                    tracing::error!("Error in job: {e}");
                    let error = gadget_sdk::events_watcher::Error::Handler(Box::new(e));
                    return Err(error);
                }
            };
        }
    } else {
        quote! {
            let job_result = match #fn_name(
                #(#additional_params_in_call)*
                #(#fn_call_params)*
            ) {
                Ok(r) => r,
                Err(e) => {
                    tracing::error!("Error in job: {e}");
                    let error = gadget_sdk::events_watcher::Error::Handler(Box::new(e));
                    return Err(error);
                }
            };
        }
    };

    let result_tokens = if result.len() == 1 {
        let ident = format_ident!("job_result");
        match event_handler {
            EventHandlerArgs::Eigenlayer { .. } => {
                vec![quote! { let #ident = job_result; }]
            }
            EventHandlerArgs::Tangle => {
                vec![crate::tangle::field_type_to_result_token(
                    &ident, &result[0],
                )]
            }
        }
    } else {
        result
            .iter()
            .enumerate()
            .map(|(i, t)| {
                let ident = format_ident!("result_{i}");
                match event_handler {
                    EventHandlerArgs::Eigenlayer { .. } => {
                        quote! {
                            let #ident = job_result[#i];
                        }
                    }
                    EventHandlerArgs::Tangle => {
                        let s = crate::tangle::field_type_to_result_token(&ident, t);
                        quote! {
                            let #ident = job_result[#i];
                            #s
                        }
                    }
                }
            })
            .collect::<Vec<_>>()
    };

    if event_handler.is_eigenlayer() {
        generate_eigenlayer_event_handler(
            &fn_name_string,
            &struct_name,
            event_handler,
            &params_tokens,
            &additional_params,
            &fn_call,
        )
    } else {
        generate_tangle_event_handler(
            &fn_name_string,
            &struct_name,
            job_id,
            &params_tokens,
            &result_tokens,
            &additional_params,
            &fn_call,
        )
    }
}

/// `JobArgs` type to handle parsing of attributes
pub(crate) struct JobArgs {
    /// Unique identifier for the job in the blueprint
    /// `#[job(id = 1)]`
    id: LitInt,
    /// List of parameters for the job, in order.
    /// `#[job(params(a, b, c))]`
    params: Vec<Ident>,
    /// List of return types for the job, could be infered from the function return type.
    /// `#[job(result(u32, u64))]`
    /// `#[job(result(_))]`
    result: ResultsKind,
    /// Optional: Verifier for the job result, currently only supports EVM verifier.
    /// `#[job(verifier(evm = "MyVerifierContract"))]`
    verifier: Verifier,
    /// Optional: Event handler type for the job.
    /// `#[job(event_handler = "tangle")]`
    event_handler: EventHandlerArgs,
    /// Optional: Skip code generation for this job.
    /// `#[job(skip_codegen)]`
    /// this is useful if the developer want to impl a custom event handler
    /// for this job.
    skip_codegen: bool,
}

impl Parse for JobArgs {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let mut params = Vec::new();
        let mut result = None;
        let mut id = None;
        let mut verifier = Verifier::None;
        let mut event_handler = EventHandlerArgs::Tangle;
        let mut skip_codegen = false;

        while !input.is_empty() {
            let lookahead = input.lookahead1();
            if lookahead.peek(kw::id) {
                let _ = input.parse::<kw::id>()?;
                let _ = input.parse::<Token![=]>()?;
                id = Some(input.parse()?);
            } else if lookahead.peek(kw::params) {
                let Params(p) = input.parse()?;
                params = p;
            } else if lookahead.peek(kw::result) {
                let Results(r) = input.parse()?;
                result = Some(r);
            } else if lookahead.peek(kw::verifier) {
                verifier = input.parse()?;
            } else if lookahead.peek(kw::event_handler) {
                event_handler = input.parse()?;
            } else if lookahead.peek(kw::skip_codegen) {
                let _ = input.parse::<kw::skip_codegen>()?;
                skip_codegen = true;
            } else if lookahead.peek(Token![,]) {
                let _ = input.parse::<Token![,]>()?;
            } else {
                return Err(lookahead.error());
            }
        }

        let id = id.ok_or_else(|| input.error("Missing `id` argument in attribute"))?;

        if params.is_empty() {
            return Err(input.error("Missing `params` argument in attribute"));
        }

        let result = result.ok_or_else(|| input.error("Missing 'result' argument in attribute"))?;

        if let ResultsKind::Types(ref r) = result {
            if r.is_empty() {
                return Err(input.error("Expected at least one parameter for the `result` attribute, or `_` to infer the type"));
            }
        }

        Ok(JobArgs {
            id,
            params,
            result,
            verifier,
            event_handler,
            skip_codegen,
        })
    }
}

#[derive(Debug)]
pub struct Params(pub Vec<Ident>);

impl Parse for Params {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let _ = input.parse::<kw::params>();
        let content;
        let _ = syn::parenthesized!(content in input);
        let names = content.parse_terminated(Ident::parse_any, Token![,])?;
        let mut items = HashSet::new();
        let mut args = Vec::new();
        for name in names {
            if items.contains(&name) {
                return Err(syn::Error::new(
                    name.span(),
                    "tried to add the same field twice",
                ));
            }

            let inserted = items.insert(name.clone());
            assert!(inserted, "tried to add the same field twice");
            args.push(name);
        }
        Ok(Self(args))
    }
}

impl JobArgs {
    fn params_to_field_types(
        &self,
        param_types: &BTreeMap<Ident, Type>,
    ) -> syn::Result<Vec<FieldType>> {
        let params = self
            .params
            .iter()
            .map(|ident| {
                param_types.get(ident).ok_or_else(|| {
                    syn::Error::new_spanned(ident, "parameter not declared in the function")
                })
            })
            .map(|ty| type_to_field_type(ty?))
            .collect::<syn::Result<Vec<_>>>()?;
        Ok(params)
    }

    fn result_to_field_types(&self, result: &Type) -> syn::Result<Vec<FieldType>> {
        match &self.result {
            ResultsKind::Infered => type_to_field_type(result).map(|x| vec![x]),
            ResultsKind::Types(types) => {
                let xs = types
                    .iter()
                    .map(type_to_field_type)
                    .collect::<syn::Result<Vec<_>>>()?;
                Ok(xs)
            }
        }
    }
}
pub enum ResultsKind {
    Infered,
    Types(Vec<Type>),
}

impl std::fmt::Debug for ResultsKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Infered => write!(f, "Infered"),
            Self::Types(_) => write!(f, "Types"),
        }
    }
}

#[derive(Debug)]
struct Results(ResultsKind);

impl Parse for Results {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let _ = input.parse::<kw::result>();
        let content;
        let _ = syn::parenthesized!(content in input);
        let names = content.parse_terminated(Type::parse, Token![,])?;
        if names.is_empty() {
            return Err(syn::Error::new_spanned(
                names,
                "Expected at least one parameter",
            ));
        }
        if names.iter().any(|ty| matches!(ty, Type::Infer(_))) {
            // Infer the types from the retun type
            return Ok(Self(ResultsKind::Infered));
        }
        let mut items = Vec::new();
        for name in names {
            items.push(name);
        }
        Ok(Self(ResultsKind::Types(items)))
    }
}

#[derive(Debug)]
enum Verifier {
    None,
    /// #[job(verifier(evm = "`MyVerifierContract`"))]
    Evm(String),
}

impl Parse for Verifier {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let _ = input.parse::<kw::verifier>()?;
        let content;
        let _ = syn::parenthesized!(content in input);
        let lookahead = content.lookahead1();
        // parse `(evm = "MyVerifierContract")`
        if lookahead.peek(kw::evm) {
            let _ = content.parse::<kw::evm>()?;
            let _ = content.parse::<Token![=]>()?;
            let contract = content.parse::<LitStr>()?;
            Ok(Verifier::Evm(contract.value()))
        } else {
            Ok(Verifier::None)
        }
    }
}

pub(crate) enum EventHandlerArgs {
    Tangle,
    Eigenlayer {
        instance: Option<Ident>,
        event: Option<Type>,
        event_converter: Option<Type>,
        callback: Option<Type>,
    },
}

impl EventHandlerArgs {
    pub fn is_eigenlayer(&self) -> bool {
        matches!(self, Self::Eigenlayer { .. })
    }

    pub fn instance(&self) -> Option<Ident> {
        match self {
            Self::Eigenlayer { instance, .. } => instance.clone(),
            Self::Tangle => None,
        }
    }

    pub fn event(&self) -> Option<Type> {
        match self {
            Self::Eigenlayer { event, .. } => event.clone(),
            Self::Tangle => None,
        }
    }

    pub fn event_converter(&self) -> Option<Type> {
        match self {
            Self::Eigenlayer {
                event_converter, ..
            } => event_converter.clone(),
            Self::Tangle => None,
        }
    }

    pub fn callback(&self) -> Option<Type> {
        match self {
            Self::Eigenlayer { callback, .. } => callback.clone(),
            Self::Tangle => None,
        }
    }
}

impl Parse for EventHandlerArgs {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let _ = input.parse::<kw::event_handler>()?;
        let content;
        syn::parenthesized!(content in input);

        let protocol = if content.peek(kw::protocol) {
            let _ = content.parse::<kw::protocol>()?;
            let _ = content.parse::<Token![=]>()?;
            content.parse::<LitStr>()?.value()
        } else {
            "tangle".to_string()
        };

        match protocol.as_str() {
            "tangle" => Ok(EventHandlerArgs::Tangle),
            "eigenlayer" => {
                let mut instance = None;
                let mut event = None;
                let mut event_converter = None;
                let mut callback = None;

                while !content.is_empty() {
                    if content.peek(kw::instance) {
                        let _ = content.parse::<kw::instance>()?;
                        let _ = content.parse::<Token![=]>()?;
                        instance = Some(content.parse::<Ident>()?);
                    } else if content.peek(kw::event) {
                        let _ = content.parse::<kw::event>()?;
                        let _ = content.parse::<Token![=]>()?;
                        event = Some(content.parse::<Type>()?);
                    } else if content.peek(kw::event_converter) {
                        let _ = content.parse::<kw::event_converter>()?;
                        let _ = content.parse::<Token![=]>()?;
                        event_converter = Some(content.parse::<Type>()?);
                    } else if content.peek(kw::callback) {
                        let _ = content.parse::<kw::callback>()?;
                        let _ = content.parse::<Token![=]>()?;
                        callback = Some(content.parse::<Type>()?);
                    } else if content.peek(Token![,]) {
                        let _ = content.parse::<Token![,]>()?;
                    } else {
                        return Err(content.error("Unexpected token"));
                    }
                }

                Ok(EventHandlerArgs::Eigenlayer {
                    instance,
                    event,
                    event_converter,
                    callback,
                })
            }
            _ => Err(syn::Error::new_spanned(
                protocol,
                "Expected `tangle` or `eigenlayer` as event handler protocol",
            )),
        }
    }
}
