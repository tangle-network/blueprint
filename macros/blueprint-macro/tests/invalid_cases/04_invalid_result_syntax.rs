use gadget_blueprint_macro::job;

struct Bytes;
impl Bytes {
    fn new() -> Self {
        Self
    }
}

#[job(params(n), result(Bytes String))]
fn keygen(n: u16) -> (Bytes, String) {
    (Bytes::new(), String::new())
}

fn main() {}
