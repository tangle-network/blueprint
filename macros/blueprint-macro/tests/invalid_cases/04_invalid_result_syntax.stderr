error: expected `,`
  --> tests/invalid_cases/04_invalid_result_syntax.rs:10:31
   |
10 | #[job(params(n), result(Bytes String))]
   |                               ^^^^^^
